### ISSUE TITLE
--kernel-memory not recognizing anon pages

### ISSUE THREAD

**Issue by AlexChung1995** [distribution/distribution#3504]
Hello, 
hoping you could help me understand how docker monitors kernel memory usage and kills container when that limit is reached.

We have the following go program running in a container based on alpine:3.12

`package main
import (
        "log"
        "time"

        "github.com/google/gopacket"
        "github.com/google/gopacket/pcap"
)
func main() {
        log.Println("starting server")
        const limit = 100
        packetSources := make([]*gopacket.PacketSource, limit)
        handles := make([]*pcap.Handle, limit)
        deviceName := "any"
        for i := 0; i < limit; i++ {
                inactive, err := pcap.NewInactiveHandle(deviceName)
                if err != nil {
                        log.Println("couldnt open inactive")
                        continue
                }
                defer inactive.CleanUp()
                if err := inactive.SetImmediateMode(true); err != nil {
                        log.Println("couldnt set immediate")
                        continue
                } else if err := inactive.SetSnapLen(4800); err != nil {
                        log.Println("couldnt set snaplen")
                        continue
                } else if err := inactive.SetBufferSize(16 * 1024 * 1024); err != nil {
                        log.Println("couldnt set buffer size")
                        continue
                } else if err := inactive.SetTimeout(time.Millisecond * time.Duration(100)); err != nil { // set negative timeout, mechanics described here: https://godoc.org/github.com/google/gopacket/pcap#hdr-PCAP_Timeouts
                        log.Println("couldnt set timeout")
                        continue
                }
                handle, err := inactive.Activate()
                if err != nil {
                        log.Println("couldnt activate")
                        continue
                }
                handles[i] = handle
                err = handle.SetBPFFilter("icmp")
                packetSources[i] = gopacket.NewPacketSource(handle, handle.LinkType())
        }
        time.Sleep(30 * time.Second)
        for _, hand := range handles {
                hand.Close()
        }
}`

Basically this program opens n packet capture handles by using the pcap library. You can obtain the pcap lib with "apk update && apk add libpcap" in the container.

When we run the container, it consumes increasing amounts of memory, far beyond the kernel memory limit and regular memory limit we set.

We've found that these pcap handles appear to be memory owned by anon pages and their usage can be tracked in /proc/<pid>/smaps where <pid> is the process id of our container. 

Could you please share some info on how docker manages kernel memory usage? How can we protect physical resources from memory leaks like this?

**Comment by milosgajdos** [comment]
Please open an issue on Moby repo. This is a distribution aka registry repo.
